macros/src/main/scala/shine/macros/Primitive.scala

```diff
@@ -0,0 +1,416 @@
+              case AppliedTypeTree((Ident(TypeName("Vector")),
+              List(AppliedTypeTree((Ident(TypeName("Phrase")), _)))))
+                   |   AppliedTypeTree((Ident(TypeName("Seq")),
+              List(AppliedTypeTree((Ident(TypeName("Phrase")), _)))))
+              =>
+                q"$name.flatMap(shine.DPIA.Phrases.xmlPrinter(_)):_*"
+              case _ =>
+                q"scala.xml.Text(shine.DPIA.Phrases.ToString($name))"
+            }
+            q"""
+               scala.xml.Elem(null, ${name.toString},
+                              scala.xml.Null, scala.xml.TopScope,
+                              minimizeEmpty = false, $body)
+             """
+        }
+      }
+
+      val lowerCaseName = makeLowerCaseName(name.toString)
+      val (rest, attributes) = makeAttributes(params)
+      val body = makeBody(rest)
+
+      q"""
+       override def xmlPrinter: scala.xml.Elem = {
+         val attributes_ = $attributes
+         val body_ = $body
+         scala.xml.Elem(null, $lowerCaseName, attributes_, scala.xml.TopScope,
+                        minimizeEmpty = false, (body_):_*)
+       }
+       """
+    }
+
+    def makeExpPrimitiveClass(name: TypeName,
+                              additionalParams: List[ValDef],
+                              params: List[ValDef],
+                              body: List[Tree],
+                              parents: List[Tree]): ClassDef = {
+      var visitAndRebuildMissing = true
+      var xmlPrinterMissing = true
+      body.foreach {
+        case DefDef(_, TermName("visitAndRebuild"), _, _, _, _) =>
+          visitAndRebuildMissing = false
+        case DefDef(_, TermName("xmlPrinter"), _, _, _, _) =>
+          xmlPrinterMissing = false
+        case _ =>
+      }
+
+      val generated = q"""
+          ${if (visitAndRebuildMissing)
+        makeVisitAndRebuild(name, additionalParams, params)
+      else q""}
+
+          ${if (xmlPrinterMissing)
+        makeXMLPrinter(name, additionalParams, params)
+      else q""}
+         """
+
+      val expClass = (additionalParams match {
+        case List() =>
+          q"""
+          final case class $name(..$params) extends {} with ..$parents {
+            ..$body
+            ..$generated
+          }
+         """
+        case _ =>
+          val newParams = params map {
+            case ValDef(_, name, tpt, rhs) if rhs.isEmpty => q"val $name : $tpt"
+            case ValDef(_, name, tpt, rhs) => q"val $name : $tpt = $rhs"
+          }
+          q"""
+          final case class $name(..$additionalParams)
+                                (..$newParams) extends {} with ..$parents {
+            ..$body
+            ..$generated
+          }
+         """
+      }).asInstanceOf[ClassDef]
+
+      c.info(c.enclosingPosition,
+        s"generated `${name.toString}'\n$expClass", force = false)
+
+      expClass
+    }
+
+    def expPrimitivesFromClassDef: ClassDef => ClassDef = {
```

Comment:
Would it be possible to have a single `*PrimitivesFromClassDef` that accepts the `make*PrimitiveClass` as an argument?